apps/test-app/frontend/src/components/tree-widget/stateless/Tree.tsx

```diff
@@ -107,24 +108,59 @@ export function StatelessTreeWidget(props: Omit<TreeWidgetProps, "rulesetId">) {
     };
   }, [modelsTreeHierarchyProvider, hierarchyLevelSizeLimit]);
 
-  const modelSource = useTreeModelSource(dataProvider);
-  const treeModel = useTreeModel(modelSource);
-  const nodeLoader = useTreeNodeLoader(dataProvider, modelSource);
+  const { componentsState, onReload } = useControlledTreeComponentsState(dataProvider);
+  const treeModel = useTreeModel(componentsState.modelSource);
 
   const [eventHandler, setEventHandler] = useState<TreeEventHandler>();
   useEffect(() => {
-    const handler = new UnifiedSelectionTreeEventHandler({ imodel: props.imodel, nodeLoader, collapsedChildrenDisposalEnabled: true });
+    const handler = new UnifiedSelectionTreeEventHandler({
+      imodel: props.imodel,
+      nodeLoader: componentsState.nodeLoader,
+      collapsedChildrenDisposalEnabled: true,
+    });
     setEventHandler(handler);
     return () => handler.dispose();
-  }, [props.imodel, nodeLoader, modelSource]);
+  }, [props.imodel, componentsState.nodeLoader, componentsState.modelSource]);
 
   const nodeRenderer = (nodeProps: TreeNodeRendererProps) => (
     <StatelessTreeNodeRenderer
       {...nodeProps}
       onLimitReset={(parentId?: string) => setHierarchyLevelSizeLimit((map) => ({ ...map, [parentId ?? ""]: "unbounded" }))}
     />
   );
+
+  const [shouldUseCustomFormatter, setShouldUseCustomFormatter] = useState<boolean>(false);
+  const onItemsRendered = useFormatter({
+    formatter: shouldUseCustomFormatter ? customFormatter : undefined,
+    dataProvider,
+    modelSource: componentsState.modelSource,
+    modelsTreeHierarchyProvider,
+    onReload,
+  });
+
   const { headerRef, treeHeight } = useTreeHeight(props.height);
+  const dropdownMenuItems = (close: () => void) => [
+    <MenuItem
+      key={1}
+      onClick={() => {
+        setShouldUseCustomFormatter(true);
+        close();
+      }}
+      style={shouldUseCustomFormatter ? { display: "none" } : {}}
```

Comment:
Maybe do not add `MenuItem` to the list instead of hiding it through `display: "none"`

Reply:
Using single `MenuItem` now